test/ssl_handshake_rtt_test.c

```diff
@@ -0,0 +1,144 @@
+ * Test 1: Serverside handshake RTT (TLSv1.2)
+ * Test 2: Clientside handshake RTT (TLSv1.3)
+ * Test 3: Serverside handshake RTT (TLSv1.3)
+ * Test 4: Clientside handshake RTT with Early Data (TLSv1.3)
+ */
+static int test_handshake_rtt(int tst)
+{
+    SSL_CTX *cctx = NULL, *sctx = NULL;
+    SSL *clientssl = NULL, *serverssl = NULL;
+    int testresult = 0;
+    SSL_CONNECTION *s = NULL;
+    OSSL_STATEM *st = NULL;
+    uint64_t rtt;
+
+#ifdef OPENSSL_NO_TLS1_2
+    if (tst <= 1)
+        return 1;
+#endif
+#ifdef OSSL_NO_USABLE_TLS1_3
+    if (tst >= 2)
+        return 1;
+#endif
+
+    if (!TEST_true(create_ssl_ctx_pair(libctx, TLS_server_method(),
+                                       TLS_client_method(),
+                                       TLS1_VERSION,
+                                       (tst <= 1) ? TLS1_2_VERSION
+                                                  : TLS1_3_VERSION,
+                                       &sctx, &cctx, cert, privkey))
+            || !TEST_true(create_ssl_objects(sctx, cctx, &serverssl, &clientssl,
+                                             NULL, NULL)))
+        goto end;
+
+    /* set client or server ssl object as purpose of test */
+    if (tst % 2 == 0) {
+        s = SSL_CONNECTION_FROM_SSL(clientssl);
+        st = &s->statem;
+    } else {
+        s = SSL_CONNECTION_FROM_SSL(serverssl);
+        st = &s->statem;
+    }
+    if (s == NULL || st == NULL) return 0;
+
+    /* implicitly set handshake rtt with a delay */
+    switch (tst) {
+        case 0:
+            st->hand_state = TLS_ST_CW_CLNT_HELLO;
+            ossl_statem_client_write_transition(s);
+            OSSL_sleep(1);
+            st->hand_state = TLS_ST_CR_SRVR_DONE;
+            ossl_statem_client_write_transition(s);
+            break;
+        case 1:
+            st->hand_state = TLS_ST_SW_SRVR_DONE;
+            ossl_statem_server_write_transition(s);
+            OSSL_sleep(1);
+            st->hand_state = TLS_ST_SR_FINISHED;
+            ossl_statem_server_write_transition(s);
+            break;
+        case 2:
+            st->hand_state = TLS_ST_CW_CLNT_HELLO;
+            ossl_statem_client_write_transition(s);
+            OSSL_sleep(1);
+            st->hand_state = TLS_ST_CR_SRVR_DONE;
+            ossl_statem_client_write_transition(s);
+            break;
+        case 3:
+            st->hand_state = TLS_ST_SW_SRVR_DONE;
+            ossl_statem_server_write_transition(s);
+            OSSL_sleep(1);
+            st->hand_state = TLS_ST_SR_FINISHED;
+            ossl_statem_server_write_transition(s);
+            break;
+        case 4:
+            st->hand_state = TLS_ST_EARLY_DATA;
+            ossl_statem_client_write_transition(s);
+            OSSL_sleep(1);
+            st->hand_state = TLS_ST_CR_SRVR_DONE;
+            ossl_statem_client_write_transition(s);
+            break;
+    }
+
+    if (!SSL_get_handshake_rtt((const SSL*) s, &rtt)) 
```

Comment:
I'm confused why you first use `SSL_CONNECTION_FROM_SSL`
and then cast the result back to an SSL pointer.
I the cast really makes sense, please adapt spacing to `(const SSL *)s`

Reply:
Even better: `SSL_CONNECTION_GET_SSL(s)`